### ISSUE TITLE
Security and privacy targets not defined

### ISSUE THREAD

**Issue by sakimura** [oauth-wg/oauth-selective-disclosure-jwt#87]
# Comments
From the current version of the document, the security and privacy target is not really defined but only a mechanism is defined as “An SD-JWT, at its core, is a digitally signed document containing hashes over the claim values with unique salts and other metadata. It MUST be digitally signed using the issuer's private key”. Explicitly defining security and privacy goals may be desired. 

# Proposal
 
Add the following at the beginning of “Concept”

From the point of view of the collection limitation, which is one of the main privacy principles, it is desirable to be able to provide only the claims necessary for the verifier to process. In the case of a standard JWT [RFC7519], however, this can only be achieved when the JWT is minted afresh each time and it is not possible to re-use the JWT that is pre-minted for other occasions. For example, a holder may want to re-use the document that contains name, address (municipality, street address, room number), date of birth, and National Identification Number to prove that he is a resident of the municipality so that he can receive the service to the resident. In the case of the normal JWT, this is not possible as the signature is over the entire claim set directly. To achieve it, a JWT that only contains the claim about the municipality needs to be minted. 

Selective disclosure JWT (SD-JWT) aims to address this issue by introducing an indirection: sign the salted hash of the claims instead of the claim set and provide only the claim value of the claim that is required to be processed by the receiver, which assumes the verifier role. This way, while the claim names and the salted hash of the claim values of all the claims contained in the document issued by the issuer is disclosed, the claim values of other claims are not in principle. 

In this document, a protocol is considered to fulfil the selective disclosure property when it has the following properties: 

## R1. Strong uncoverability

A protocol is said to achieve strong uncoverability if the adversary acting as the verifier cannot successfully find the claim values that were not intended to be disclosed to the verifier, with a probability significantly better than negligible. 

## R2. Weak uncoverability

While strong uncoverability is desirable, it is not trivial to achieve it. For example, if the original document was a passport that came with the signature of the national authority, then just by looking at the issuer and its signature, it is trivial to find the nationality of the subject in the release. Also, if the existence of the claim name in the release makes it possible for the verifier to guess the value with a probability significantly better than negligible, then it does not satisfy strong uncoverability. 

With these considerations, weak uncoverability is defined as follows: 

A protocol is said to achieve weak uncoverability if the adversary acting as the verifier cannot successfully find the claim values that were not intended to be disclosed to the verifier from the section of the release that directly corresponds with the claim, with a probability significantly better than negligible.  

## R3. Release confidentiality

A protocol is said to achieve release confidentiality if the adversary that can observe the content of all protocol exchanges between the holder agent and the verifier cannot find the claims with a probability significantly better than negligible. 

## R4. Unforgeability

Unforegeability ensures that an adversary acting as a holder (dishonest holder) cannot forge a release that makes the verifier believe the false value.  

The game is defined as follows. Arbitrary entities play the role of issuer and verifier while the adversary plays the role of the holder at each stage of the protocol. A is subjected to the game where sequentially: 

1. Issuer and verifier execute the setup/registration phase (if any). 
2. Issuer and adversary playing the role of holder execute the issuance phase in which issuer issues claim set attached to the honest holder C to the adversary. 
3. Adversary playing the role of holder and verifier executes the release phase for a policy P that is not satisfied by C. 
4. The winning event for the adversary is defined as a successful authentication towards the verifier. 

The protocol achieves unforgeability if it is unfeasible for the adversary to win with a probability significantly better than negligible. 

## Security and privacy target of SD-JWT and SD-JWT release

T1. SD-JWT, SD-JWT release with associated protocol execution fulfils R2, R3, and R4. 

Note: See privacy consideration for more discussion.

**Comment by peppelinux** [comment]
Would R1 be covered by this proposal https://github.com/oauthstuff/draft-selective-disclosure-jwt/issues/80 ?

**Comment by sakimura** [comment]
@peppelinux I will review it tomorrow. I took a sleeping pill a few minutes ago and probably not good to review it in this state.  (It is 1:17 AM now.) 

By the way, the current draft does not even fulfil R2 except in special cases where there are large enough value possibilities in the value itself. Is this really what we want? If this is what the community wants, I am ok with it, but my preference is to make it a bit better to require R2. 

In the privacy consideration, PR #91, I have provided one way of doing it. when the value is supposed to be text. 
Basically, making it as follows work. 

SD-CLAIMS = (
    CLAIM-NAME: HASH(SALT | HASH(SALT | CLAIM-VALUE))
)*

**Comment by Sakurann** [comment]
In my understanding, we are not defining a protocol. we are defining an extension to JWT, which is defines as a "compact, URL-safe means of representing claims to be transferred between two parties" over the protocol of the choice of those parties.

Which is why I am also not comfortable defining properties R1-R4 in this document. For example R3 is out of scope of this specification defining means of representing claims, and is in scope for a specification defining "protocol exchanges". We could add text in privacy considerations section guiding implementers to consider these aspects when choosing the "exchange protocol". Are notions R1-R4 defined in any other document? is there are a reference we can point to?

**Comment by sakimura** [comment]
Actually, I re-read the draft and it probably does provide R2. Anyway, we need to decide on what we want to achieve and check that it indeed does.

**Comment by danielfett** [comment]
I like Nat's proposal to define these or similar terms in our document. A lot!

What we claim to achieve in the document needs to be guided by the requirements, and I think we need to discuss this in more details (e.g., hiding claim names).

@Sakurann 
> In my understanding, we are not defining a protocol

We very much are, just not a network or web protocol. What is described in Concepts is essentially a protocol, without calling it that way - there are three parties and they exchange some messages. Nat's proposed adversary fits right into this.

The assumptions we make about this protocol (participants, adversaries) and the goals we want to achieve must guide our design process. This has been implicit so far, we should make it more explicit.

**Comment by sakimura** [comment]
> Would R1 be covered by this proposal #80 ?

Yes, it goes towards it, although not completely. As described in my proposed privacy consideration text, the issuer identifier alone may be enough to break this so we may need to combine it with a group signature scheme or something like that.

**Comment by Sakurann** [comment]
So in layman terms, what is being proposed in this PR and the issue are the following security goals, correct?
1. SD-JWT does not allow the attacker to find the claim values that were intended to be disclosed to another legitimate verifier.
2. SD-JWT does not allow the attacker to obtain claims even if it has access to all of the protocol exchanges between the holder and the legitimate verifier.
3. SD-JWT does not allow the attacker to create a release that would make verifier to trust claim values not originally signed by the issuer.
 
   - On 1 and 2, this is where in my mind this cannot be achieved only by using SD-JWT, but requires other "protocols". Plain text release contains actual claim values! so to achieve this property, release has to be sent over TLS, or session encrypted BLE/NFC, but we cannot force that in this SD-JWT document itself. and this is where audience restriction and nonces also come into play.
   - On 3, I agree with this one. it is achieved by including hashes in SD-JWT, undermined if verifier does not perform hash computation.

**Comment by Sakurann** [comment]
I agree we need to explain security goals of the design choices made in SD-JWT document. Could we first agree on those security goals? 

To those three outlined in this document above, I would add basic ones:
- Integrity protection - Issuer's signature on SD-JWT
- optionally, when holder binding is present - prevention of cloning of SD-JWT and MITM by signing the release using holder's key

I am sure I missed something, and maybe this has been meant in current PR text, but I re-read the PR text couple of times and was still left confused..

**Comment by Sakurann** [comment]
Once we agree on the security goals, could we also brainstorm fitting terms - "strong uncoverability", "weak uncoverability", "release confidentiality", "unforgeability" have not been the most intuitive for me, but maybe that's only me.

**Comment by danielfett** [comment]
We merged PR #225 to define the security targets. I'll keep this issue open and assign to me to define adversary capabilities in more detail for one of the next draft versions.
